components/ledger/ledger-common-flow/src/main/kotlin/net/corda/ledger/common/flow/impl/transaction/TransactionSignatureServiceImpl.kt

```diff
@@ -65,80 +84,178 @@ class TransactionSignatureServiceImpl @Activate constructor(
+        }
+        require(transactions.map { it.batchMerkleTreeDigestOptionsNodePrefix }.distinct().size == 1) {
+            "Notary merkle tree digest node prefixes should be the same in a batch to be signed."
+        }
+        require(transactions.none {
+            it.batchMerkleTreeDigestOptionsLeafPrefix contentEquals it.batchMerkleTreeDigestOptionsNodePrefix
+        }) {
+            "Notary merkle tree digest node prefixes and leaf prefixes need to be different for each transactions."
+        }
+
+        // TODO CORE-6615 This check needs to be reconsidered in the future
+        val algorithms = transactions.map { it.id.algorithm }.distinct()
         require(algorithms.size == 1) {
             "Cannot sign a batch with multiple hash algorithms: $algorithms"
         }
+        confirmHashPrefixesAreDifferent(transactions)
+    }
 
-        val algorithm = algorithms.first()
-
-        val allLeaves = txIds.map {
-            // we don't have a reHash function anymore
-            digestService.hash(it.bytes, DigestAlgorithmName(algorithm))
+    private fun confirmHashPrefixesAreDifferent(transactions: List<TransactionWithMetadata>) {
+        require(transactions.none { it.batchMerkleTreeDigestOptionsLeafPrefix contentEquals it.rootMerkleTreeDigestOptionsLeafPrefix }) {
+            "The transaction can be batch signed only if its root and the notary leaf prefixes are different."
         }
+        require(transactions.none { it.batchMerkleTreeDigestOptionsNodePrefix contentEquals it.rootMerkleTreeDigestOptionsNodePrefix }) {
+            "The transaction can be batch signed only if its root and the notary node prefixes are different."
+        }
+    }
 
-        val hashDigest = merkleTreeProvider.createHashDigestProvider(
-            HASH_DIGEST_PROVIDER_DEFAULT_NAME,
-            DigestAlgorithmName(algorithm)
-        )
+    @Suspendable
+    private fun getAvailableKeysFor(publicKeys: Iterable<PublicKey>): List<PublicKey> {
+        val availableKeys = signingService.findMySigningKeys(publicKeys.toSet()).values.filterNotNull()
+        if (availableKeys.isEmpty()) {
+            throw (TransactionNoAvailableKeysException(
+                "The publicKeys do not have any private counterparts available.",
+                null
+            ))
+        }
+        return availableKeys
+    }
 
-        val merkleTree = merkleTreeProvider.createTree(
-            allLeaves.map { it.bytes },
-            hashDigest
-        )
-        // End of Copy
+    private fun confirmSignatureBatchMerkleSettingsMatchWithTransactions(
+        transaction: TransactionWithMetadata,
+        signatureWithMetadata: DigitalSignatureAndMetadata
+    ) {
+        require(transaction.batchMerkleTreeDigestProviderName == signatureWithMetadata.batchMerkleTreeDigestProviderName) {
+            "Batch signature's digest provider should match with the transaction's notary merkle digest provider."
```

Comment:
Just a stylistic comment - I would drop all the `'s` from the messages... 
```suggestion
            "Batch signature digest provider should match with the transaction notary merkle digest provider."
```

Reply:
Thanks.
I have removed them.